controllers/cloud.redhat.com/providers/web/impl.go

```diff
@@ -68,17 +91,659 @@ func (web *webProvider) makeService(deployment *crd.Deployment, app *crd.ClowdAp
 		)
 	}
 
-	utils.MakeService(s, nn, map[string]string{"pod": nn.Name}, servicePorts, app, web.Env.IsNodePort())
+	utils.MakeService(s, nn, map[string]string{"pod": nn.Name}, servicePorts, app, env.IsNodePort())
 
 	d.Spec.Template.Spec.Containers[0].Ports = containerPorts
 
-	if err := web.Cache.Update(CoreService, s); err != nil {
+	if err := cache.Update(CoreService, s); err != nil {
 		return err
 	}
 
-	if err := web.Cache.Update(deployProvider.CoreDeployment, d); err != nil {
+	if err := cache.Update(deployProvider.CoreDeployment, d); err != nil {
 		return err
 	}
 
 	return nil
 }
+
+func makeKeycloak(o obj.ClowdObject, objMap providers.ObjectMap, usePVC bool, nodePort bool) {
+	nn := providers.GetNamespacedName(o, "keycloak")
+
+	dd := objMap[WebKeycloakDeployment].(*apps.Deployment)
+	svc := objMap[WebKeycloakService].(*core.Service)
+
+	labels := o.GetLabels()
+	labels["env-app"] = nn.Name
+
+	labeler := utils.MakeLabeler(nn, labels, o)
+
+	labeler(dd)
+
+	replicas := int32(1)
+
+	dd.Spec.Replicas = &replicas
+	dd.Spec.Selector = &metav1.LabelSelector{MatchLabels: labels}
+
+	dd.Spec.Template.ObjectMeta.Labels = labels
+
+	dd.Spec.Template.Spec.ImagePullSecrets = []core.LocalObjectReference{{
+		Name: "quay-cloudservices-pull",
+	}}
+
+	// get the secret
+
+	port := int32(8080)
+
+	envVars := []core.EnvVar{
+		{
+			Name:  "DB_VENDOR",
+			Value: "h2",
+		},
+		{
+			Name:  "PROXY_ADDRESS_FORWARDING",
+			Value: "true",
+		},
+		{
+			Name:  "KEYCLOAK_USER",
+			Value: "admin",
+		},
+		{
+			Name:  "KEYCLOAK_PASSWORD",
+			Value: "admin",
+		},
+	}
+
+	ports := []core.ContainerPort{{
+		Name:          "service",
+		ContainerPort: port,
+	}}
+
+	probeHandler := core.Handler{
+		TCPSocket: &core.TCPSocketAction{
+			Port: intstr.IntOrString{
+				Type:   intstr.Int,
+				IntVal: 8080,
+			},
+		},
+	}
+
+	livenessProbe := core.Probe{
+		Handler:             probeHandler,
+		InitialDelaySeconds: 10,
+		TimeoutSeconds:      2,
+	}
+	readinessProbe := core.Probe{
+		Handler:             probeHandler,
+		InitialDelaySeconds: 20,
+		TimeoutSeconds:      2,
+	}
+
+	c := core.Container{
+		Name:           nn.Name,
+		Image:          "quay.io/keycloak/keycloak:11.0.3",
+		Env:            envVars,
+		Ports:          ports,
+		LivenessProbe:  &livenessProbe,
+		ReadinessProbe: &readinessProbe,
+	}
+
+	dd.Spec.Template.Spec.Containers = []core.Container{c}
+	dd.Spec.Template.SetLabels(labels)
+
+	servicePorts := []core.ServicePort{{
+		Name:     "keycloak",
+		Port:     port,
+		Protocol: "TCP",
+	}}
+
+	utils.MakeService(svc, nn, labels, servicePorts, o, nodePort)
+
+}
+
+func makeBOP(o obj.ClowdObject, objMap providers.ObjectMap, usePVC bool, nodePort bool) {
```

Comment:
consider casing `o` to a `ClowdEnv` and then you can get rid of the last two args

Reply:
This won't get rid of the args as the app doesn't have this information - so unless we make two funcs - we still need this and right now, it's not worth it.